Signal/src/util/ImageCache.swift

```diff
@@ -0,0 +1,53 @@
+//
+//  Copyright (c) 2017 Open Whisper Systems. All rights reserved.
+//
+
+import Foundation
+import UIKit
+
+class ImageCacheRecord: NSObject {
+    var variations: [CGFloat: UIImage]
+    init(variations: [CGFloat: UIImage]) {
+        self.variations = variations
+    }
+}
+
+/**
+ * A two dimensional hash, allowing you to store variations under a single key.
+ * This is useful because we generate multiple diameters of an image, but when we 
+ * want to clear out the images for a key we want to clear out *all* variations.
+ */
+@objc
+class ImageCache: NSObject {
```

Comment:
I'd love to eventually a) write avatars to disk and b) evacuate the cache only when its contents change (E.g. if bob's avatar in device contacts change, only evacuate his avatar, not Alice's - and if alice's profile avatar changed _not_ evacuate her avatar if we're using an avatar from our device contacts for her).